[쇼핑카트 / 1단계] 로이(김덕우) 미션 제출합니다.

the9kim opened the pull request:
안녕하세요!

백엔드 크루 로이입니다.

스프링이 처음이다 보니, 개념을 정확히 이해하지 못하고 구현한 것 같아 마음에 걸립니다!

부족한 점을 알려주시면, 충분히 고민해보고 반영해보겠습니다.

감사합니다!

wugawuga committed:
docs: 기능 요구사항 작성

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: Controller 생성 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: index.html 상품 전체 조회 렌더링 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
docs: 기능 구현 목록 수정

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: admin.html 상품 조회 렌더링 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: 상품 등록 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: 상품 수정 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: 상품 삭제 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
feat: 상품 등록, 수정 검증 구현

Co-authored-by: the9kim <the9rla@gmail.com>

wugawuga committed:
style: 공백 라인 수정

Co-authored-by: the9kim <the9rla@gmail.com>

jnsorn commented on src/main/java/cart/controller/ProductApiController.java:
```diff
@@ -0,0 +1,42 @@
+        this.productService = productService;
+    }
+
+    @PostMapping("/product/insert")
```
로이는 어떤 기준으로 HTTP Method와 URI를 설계하시는지 궁금해요 🧐

the9kim replied:
사실 별다른 생각 없이 기능을 나타내는 동사로 URI를 설계했습니다.

REST API에 관한 내용을 좀 찾아보니, 

URI는 리소스를 기준으로 설계하되 기능별 구분 시 HTTPMETHOD를 최대한 활용한다는 기준을 알게됐습니다.

따라서 기본 URI를 'Products'로 설정하고, 제품 수정 컨트롤러의 메서드(POST -> PATCH)도 변경해봤습니다!

jnsorn replied:
URI가 더 명확해 진 것 같네요 👍

jnsorn commented on src/main/java/cart/controller/ProductApiController.java:
```diff
@@ -0,0 +1,42 @@
+            return ResponseEntity.badRequest().body(errors.getFieldError().getField());
+        }
+        productService.insert(productDto);
+        return ResponseEntity.ok().body("ok");
```
Body에 ok를 담는 이유는 무엇일까요? 🧐

the9kim replied:
@RestContoller 어노테이션을 사용했을 때,

String을 반환하면 해당 값이 Body에 바로 실리기 때문에 굳이 ResponseEntity로 포장하지 않아도 된다는 사실을 몰랐습니다ㅠ

String으로 "ok"값을 반환하는 방식으로 변경해봤습니다!

jnsorn commented on src/main/java/cart/dao/ProductEntity.java:
```diff
@@ -0,0 +1,48 @@
+package cart.dao;
+
+import cart.dto.ProductDto;
+
+public class ProductEntity {
+
```
Entity가 Dao 패키지 내에 존재하네요! 로이가 생각하는 Entity의 역할은 무엇인지 궁금해요~!

the9kim replied:
Entitiy는 DB테이블과 매칭되는 클래스로 비즈니스 로직을 담을 수 있는 것으로 알고 있습니다.

로직을 포함할 수 있다는 점에서 도메인 패키지가 좀 더 어울리는 듯 합니다!

해당 클래스를 도메인 패키지로 옮겨봤습니다!

jnsorn commented on src/main/java/cart/dao/ProductEntity.java:
```diff
@@ -0,0 +1,48 @@
+        this.image = image;
+        this.price = price;
+    }
+
```
중복되는 코드를 어떻게 제거할 수 있을까요~?

jnsorn commented on README.md:
```diff
@@ -1 +1,14 @@
 # jwp-shopping-cart
+
+### 기능 요구사항
```
테스트가 하나도 없어요~ 구현한 API가 잘 동작하는지 확인할 수 있는 테스트를 추가해주세요(Controller 필수)

the9kim replied:
컨트롤러와 dao 테스트코드 추가했습니다!

앞으로 테스트에 좀 더 신경쓰겠습니다ㅠ

jnsorn commented on src/main/java/cart/controller/ProductApiController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @PostMapping("/product/insert")
+    public ResponseEntity<String> insert(@Valid @RequestBody ProductDto productDto, Errors errors) {
+        if (errors.hasErrors()) {
+            return ResponseEntity.badRequest().body(errors.getFieldError().getField());
+        }
```
@Valid를 통한 검증, 유효성 검사에 대한 예외처리 좋습니다. 👍 
다만 현재의 구조는 DTO를 요청으로 받을 때마다 위 로직이 반복될 것 같은데요, 중복을 줄여볼 수 없을까요?
(Hint. `@ ControllerAdvice `에 대해 알아보아요.

jnsorn reviewed (changes requested):
안녕하세요. 로이! 만나서 반갑습니다. 1단계 깔끔하게 잘 구현해주셨어요! 👍
바로 다음단계로 넘어가도 좋을 것 같았는데, 넘어가기 전 테스트는 있어야 할 것 같아 다시 수정 요청드렸어요.
테스트만 추가해서 바로 다시 요청주셔도 좋습니다. 😄

the9kim committed:
fix: 데이터베이스에 상품 정보 등록시 id를 반환

the9kim committed:
refactor: Dto 내 기본 생성자 추가

the9kim committed:
refactor: 테스트를 위한 DB 초기화 로직 구현

the9kim committed:
test: 등록 제품 조회 관련 통합 테스트

the9kim committed:
test: dao 테스트

the9kim committed:
refactor: API의 HTTP Method 및 URI 변경, 반환 타입 변경

1. Http Method + URI 변경
- 리소스를 기준으로 URI 설정(products)
- 기능 구분을 위한 메서드 변경(제품 수정 : post -> patch)

2. 반환 타입(ResponseBody -> String) 변경

the9kim committed:
refactor: Entity 클래스 패키지 이동 (dao -> domain)

jnsorn commented on src/test/java/cart/controller/HomeControllerTest.java:
```diff
@@ -0,0 +1,45 @@
+                .all()
+                .statusCode(HttpStatus.OK.value());
+    }
+}
```
<img width="161" alt="image" src="https://user-images.githubusercontent.com/31095063/235809313-b6ffe004-f5cd-41ed-8f22-a63752d1fa1a.png">
😢

the9kim replied:
전체적으로 공백라인을 수정해봤습니다.

디테일에 좀 더 신경쓰겠습니다!

jnsorn commented on src/test/java/cart/controller/ProductApiControllerTest.java:
```diff
@@ -0,0 +1,111 @@
+        ProductDto productDto = new ProductDto("pizza", "https://www.hmj2k.com/data/photos/20210936/art_16311398425635_31fd17.jpg", 10000);
+        String request = objectMapper.writeValueAsString(productDto);
+
+        RestAssured.given()
+                .log()
+                .all()
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .body(request)
+                .when()
+                .post("/products")
+                .then()
+                .log()
+                .all()
+                .statusCode(HttpStatus.OK.value());
+    }
```
가독성을 위해 given, when, then 기준으로 들여쓰기가 되면 좋겠네요!

the9kim replied:
아 그런 방식이 있는지 몰랐네요ㅠ

말씀하신 방식으로 수정해봤습니다.

알려주셔서 감사합니다!

jnsorn commented on src/test/java/cart/dao/ProductDaoImplTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Test
+    @DisplayName("데이터를 추가한다")
```
데이터라는 모호한 표현 대신 실제 테스트에 대한 설명이 있으면 좋을 것 같아요~

the9kim replied:
설명이 너무 막연했네요ㅠ
displayname을 좀 더 세부적으로 수정해봤습니다.

jnsorn reviewed (approved):
안녕하세요. 로이! 테스트 추가해주신것 확인했습니다.
1단계 요구사항은 충족한 것 같아 이만 머지할게요.
2단계에서 더 많은 이야기 나눠봐요~~! 🛒🛍️
